### ISSUE TITLE
Develop a system that generates various spatial SBML models using deep learning

### ISSUE THREAD

**Issue by funasoul** [nrnb/GoogleSummerOfCode#140]
### Background
[SBML(Systems Biology Markup Language)](http://sbml.org) has several extension packages to extend its capability. One of its extension is called [Spatial Processes (spatial)](http://sbml.org/Documents/Specifications/SBML_Level_3/Packages/spatial) which supports for describing processes that involve a spatial component, and describing the geometries involved. SBML spatial extension enables users to build a spatial model and run a [spatial simulation](https://github.com/funasoul/docker-spatialsim/raw/master/images/sam2d_s1_cyt.gif).

We have been working on the development of a software tool, [XitoSBML](https://github.com/spatialsimulator/XitoSBML): a spatial model builder that will generate a spatial SBML model from microscopic images. 
Although XitoSBML provides a user-friendly UI to create a spatial SBML model, there only exists few spatial SBML models for spatial simulation. In order to use XitoSBML, cell regions in microscopic images of cells need to be segmented beforehand by image processing. This process (which is called segmentation) is a bottleneck of creating spatial SBML models.
On the other hand, the accuracy of image processing using deep learning has been remarkable in recent years, and methods for [highly accurate segmentation](https://arxiv.org/abs/1505.04597) have been proposed.

This summer, we would like to mentor a student who will implement a system that automatically generates various spatial SBML models by comprehensively segmenting microscopic images of cells using deep learning and XitoSBML.

### Goal
As described in the "Background", the project could be split down into following 4 tasks.

- Collect images from a database in which microscopic images of cells have been published.
- Prepare some datasets which will be used by deep learning algorithm for performing segmentation.
- Train the learning machine (deep learning algorithm) to comprehensively segment microscopic images of cells
- Implement a software tool to automatically convert segmented images to Spatial SBML Models using XitoSBML

The following API / frameworks will be used for this project.

- [PyTorch](https://pytorch.org/) ... for deep learning
- [ImageJ API](https://imagej.net/Developing_Plugins_for_ImageJ_1.x) ... for XitoSBML
- [JSBML](http://sbml.org/Software/JSBML) ... for XitoSBML

### Difficulty Level 2
Although this project seems to have many tasks to solve, each of the tasks will not require enormous lines/time to code, because there already exists convenient API, well-documented API docs and an existing implementation of the learning machine. The most important part is to understand the specification of the SBML spatial extension and preparing the training datasets.

### Skills
Java and Python programming skills and some basic knowledge of handling XML documents are required. Nice to have knowledge/experience on SBML, image processing and mathematical background on machine learning.

- (essential) Java, Python, XML
- (nice to have) SBML, Image processing, mathematical background on machine learning

### Public Repository
- [XitoSBML](https://github.com/spatialsimulator/XitoSBML)

### References
- [SBML Spatial specification rel 0.94](https://sourceforge.net/p/sbml/code/HEAD/tree/trunk/specifications/sbml-level-3/version-1/spatial/specification/spatial-v1-sbml-l3v1-rel0.94.pdf)
- [JSBML Documentation](http://sbml.org/Software/JSBML/docs)
- [ImageJ API](https://imagej.net/Developing_Plugins_for_ImageJ_1.x)
- [PyTorch](https://pytorch.org/)

### Potential Mentors
- [Akira Funahashi](https://github.com/funasoul) Keio University, Japan
- [Yuta Tokuoka](https://github.com/tokkuman) Keio University, Japan
- [Kaito Ii](https://github.com/kaitoii11) Hewlett-Packard Japan, Ltd. , Japan

### Contact
[Akira Funahashi](mailto:funasoul@gmail.com)
[Yuta Tokuoka](mailto:tokuoka@fun.bio.keio.ac.jp)
[Kaito Ii](mailto:kaitoii1111@gmail.com)

**Comment by singh-yashwant** [comment]
Hi @funasoul 
I am interested in this project and have few queries should I email or post them here only?
Thanks, have a great day

**Comment by funasoul** [comment]
Thank you for your interest in this project. Let me clarify that as we have not yet submitted our organization application, it is not confirmed to be a GSoC 2020 project yet.
But we are welcome to answer your questions on it.
Please post here so that @kaitoii11 and @tokkuman can also reply to your question.

**Comment by singh-yashwant** [comment]
Thanks for fast reply @funasoul 
I am Yashwant Singh a Computer Science sophomore at Dayananda Sagar College of Engineering, Bangalore, India.
I found this project very interesting and aligned with my skill-set which include
- Knowledge of Python and Java programming language
- basic understanding of XML 
- Experience in working with basic machine learning and deep learning models
However I am not familiar with SBML and don't have a strong mathematical background in machine learning, I would like to work on this project if it is selected for GSoC 2020 therefore I seek your help to get me started with understanding this project, SBML and XitoSBML api.

Thanks

**Comment by born-2learn** [comment]
@funasoul @kaitoii11 @tokkuman 

Image Processing and deep learning applications have already fascinated me since high school.

I'm Syed Farhan, an Undergrad student in India, and am interested to explore this project.
I have experience with python and tensorflow, and can also try and work with pytorch.

**Comment by funasoul** [comment]
@singh-yashwant @born-2learn 
Thank you for your interest in this project.
Please send your CV to us (the 3 email addresses provided above in "Contact"). Thanks!

**Comment by born-2learn** [comment]
@funasoul 
Thank you for your response.
I have shared my resume(Cv) with all the three contacts.(from: syedfarhana.ec18@rvce.edu.in)
Looking forward to working with you.

**Comment by born-2learn** [comment]
@funasoul @kaitoii11 @tokkuman 
Can you guide me with getting started and the further steps to be taken.
Is it compulsory to use pytorch instead of tensorflow?

**Comment by funasoul** [comment]
> @funasoul @kaitoii11 @tokkuman
> Can you guide me in getting started and the further steps to be taken.
> Is it compulsory to use pytorch instead of tensorflow?

Just sent you my reply to your email address.

**Comment by ayushtues** [comment]
Hi @funasoul, I am a deep learning enthusiast and I especially love using Pytorch and have done various computer vision projects using it. Can you guide me on how to proceed further?

**Comment by funasoul** [comment]
@ayushtues 
It is great that you already have experience on machine learning and computer vision.
Please send your CV to us (the 3 email addresses provided above in "Contact"). Thanks!

**Comment by ayushtues** [comment]
@funasoul, I have sent the CV as requested

**Comment by funasoul** [comment]
@ayushtues Thanks. I will reply to you soon.

**Comment by Medha-B** [comment]
Hello @funasoul . I am very interested in this project and have prior experience with machine learning and deep learning. I have sent you a mail with my CV as instructed here. Kindly guide me through the process further.

**Comment by funasoul** [comment]
> Hello @funasoul . I am very interested in this project and have prior experience with machine learning and deep learning. I have sent you a mail with my CV as instructed here. Kindly guide me through the process further.

Hi @Medha-B . I just sent you my reply to your email address.

**Comment by ShreyanshTripathi** [comment]
Hi @funasoul. I am a third year undergraduate student from India. I have experience in implementing deep learning segmentation models using Pytorch. I am interested in taking up this project for GSoC this year. I am mailing you my CV (learning from the pattern). Please let me know about the further process.

**Comment by kritikajoshi26** [comment]
hi @funasoul . I am currently in a Mtech 3rd-year student from India. I have experience in developing deep learning models using PyTorch. i am interested in taking up this project. i am mailing you my resume. please let me know further

**Comment by khanspers** [comment]
Active project for GSoC 2020, closing here.
